naming/src/main/java/org/jboss/as/naming/InServerInitialContextFactoryService.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2022, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.naming;
+
+import java.util.ArrayList;
+import java.util.Hashtable;
+import java.util.List;
+
+import javax.naming.Context;
+import javax.naming.NamingException;
+import javax.naming.spi.InitialContextFactory;
+
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+
+/**
+ * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
+ */
+public final class InServerInitialContextFactoryService implements InitialContextFactory, Service<InitialContextFactory> {
+
+    public static final ServiceName SERVICE_NAME = ServiceName.JBOSS.append("InServerInitialContextFactoryService");
+
+    private List<LookupInterceptor> interceptors = new ArrayList<>();
+
+    public void addInterceptor(LookupInterceptor interceptor){
+        interceptors.add(interceptor);
+    }
+
+    public void removeInterceptor(LookupInterceptor interceptor){
+        interceptors.remove(interceptor);
+    }
+
+    @Override
+    public InitialContextFactory getValue() throws IllegalStateException, IllegalArgumentException {
+        return this;
+    }
+
+    @Override
+    public void start(StartContext startContext) throws StartException {
+        InitialContext.setInitialContextFactory(this);
```

Comment:
This has the effect that InitialContext has 1 delegate for a period of time and then this switches it when MSC gets around to starting this service. That's fragile.

Couldn't InitialContext expose a (package-protected) static getter instead of a setter? So it creates an InitialContextFactory like this and exposes it via the static getter. This service just makes that object available as an MSC service, for injection into TransactionNamespaceAccessControlService.

The value type for this service would be a new interface "InServerInitialContextFactory" that has the add/removeInterceptor methods.